src/inetaddr.c

```diff
@@ -513,530 +258,1438 @@ int idev_add_mcast_init(struct netif_port *dev)
+    if (!(ifa->flags & IFA_F_PERMANENT)) {
+        timeo.tv_sec = ifa->valid_lft;
+        if (is_master) {
+            dpvs_timer_cancel(&ifa->timer, true);
+            dpvs_time_rand_delay(&timeo, 1000000);
             dpvs_timer_sched(&ifa->timer, &timeo, ifa_expire, ifa, true);
         }
-    } else {
-        /* ok to cancel a timer not scheduled */
+    }
+
+    if ((ifa->af == AF_INET6) && (ifa->flags & IFA_F_PERMANENT)) {
+        ifa->flags |= IFA_F_TENTATIVE | IFA_F_OPTIMISTIC;
+        inet_ifaddr_dad_start(ifa);
+    }
+
+    RTE_LOG(DEBUG, IFA, "[%02d] %s: edit ifaddr %s\n", rte_lcore_id(), __func__,
+            inet_ntop(ifa->af, &ifa->addr, ipstr, sizeof(ipstr)));
+
+    ifa_put(ifa);
+    idev_put(idev);
+    return EDPVS_OK;
+
+errout:
+    idev_put(idev);
+    RTE_LOG(WARNING, IFA, "[%02d] %s: edit ifaddr %s failed -- %s\n", rte_lcore_id(), __func__,
+            inet_ntop(param->af, &param->addr, ipstr, sizeof(ipstr)), dpvs_strerror(err));
+    return err;
+}
+
+static int ifa_entry_del(const struct ifaddr_action *param)
+{
+    struct inet_ifaddr *ifa;
+    struct inet_device *idev;
+
+    if (!param || !param->dev || !ifa_prefix_check(param->af,
+                &param->addr, param->plen))
+        return EDPVS_INVAL;
+
+    idev = dev_get_idev(param->dev);
+    if (unlikely(!idev))
+        return EDPVS_RESOURCE;
+
+    ifa = ifa_lookup(idev, &param->addr, param->plen, param->af);
+    if (!ifa) {
+        idev_put(idev);
+        return EDPVS_NOTEXIST;
+    }
+
+    ifa_unhash(ifa);
+    ifa_put(ifa);
+
+    idev_put(idev);
+    return EDPVS_OK;
+}
+
+static int ifa_entry_flush(const struct ifaddr_action *param)
+{
+    struct inet_ifaddr *ifa, *nxt;
+    struct inet_device *idev;
+    lcoreid_t cid = rte_lcore_id();
+
+    if (!param || !param->dev)
+        return EDPVS_INVAL;
+
+    idev = dev_get_idev(param->dev);
+    if (unlikely(!idev))
+        return EDPVS_RESOURCE;
+
+    list_for_each_entry_safe(ifa, nxt, &idev->ifa_list[cid], d_list) {
+        rte_atomic32_inc(&ifa->refcnt); /* hold @ifa before unhash */
+        ifa_unhash(ifa);
+        ifa_put(ifa);
+    }
+
+    idev_put(idev);
+    return EDPVS_OK;
+}
+
+static int ifa_entry_sync(const struct ifaddr_action *param)
+{
+    struct inet_device *idev;
+    struct inet_ifaddr *ifa;
+
+    if (!param || !param->dev || !ifa_prefix_check(param->af,
+                &param->addr, param->plen))
+        return EDPVS_INVAL;
+
+    idev = dev_get_idev(param->dev);
+    if (!idev)
+        return EDPVS_RESOURCE;
+
+    ifa = ifa_lookup(idev, &param->addr, param->plen, param->af);
+    if (!ifa) {
+        idev_put(idev);
+        return EDPVS_NOTEXIST;
+    }
+
+    /* only support snyc flags now */
+    ifa->flags = param->flags;
+    if ((ifa->flags & IFA_F_DADFAILED) &&
+        (rte_lcore_id() == rte_get_master_lcore()))
+        dpvs_timer_cancel(&ifa->dad_timer, true);
+
+    ifa_put(ifa);
+    idev_put(idev);
```

Comment:
since ifa is per_core，do we still need refcnt? conn need refcnt is just because conn will use for not only one loop.

Reply:
**ifa** is referred to by **sapool**, and **sapool** holds it once resources are allocated from it.